### DISCUSSION TITLE
IS it possible to center the object?

### DISCUSSION THREAD

**Discussion by rowild** [playcanvas/supersplat#17] [Q&A]
I have played around with your fantastic platform over the last weekend and cleaned up some of my files. Then I used the "Scene Orientation" feature to place my model nicely in the center. I have then converted and loaded the file into this three.js app:
https://github.com/mkkellogg/GaussianSplats3D
But it seems that this app treats the positioning of the file completely different. 

I was wondering if there is a way to center the origin as one can do in e.g. Blender?

**Comment by slimbuck** [comment]
Hi @rowild,

What do you mean by `But it seems that this app treats the positioning of the file completely different.` ? Is the scene's origin off, or is the entire scene orientated differently too?

We adopted the scene orientation of the original gaussian splatting paper (roughly), but that's kind of arbitrary.

We should probably add two things:
- some way to specify the origin of the scene for export
- perhaps some way of specifying export "scene orientation" - "y-up", "z-up" etc

Thanks!

**Reply by slimbuck** [reply]
Different target platforms operate in different coordinate systems.

Perhaps a better approach would be to export to various target platforms once we know their details: Export orientation "Unity", "Unreal", "PlayCanvas", "Three" etc.

**Comment by rowild** [comment]
May I please describe what my workflow is – maybe my "problem" will become clearer. This is what I am doing know:
- Create and download a `.ply` in `Luma` / `poly.cam`
- Load the ply into `Super Splash` and clean up the artefacts
- Download the cleaned (Super Splash) `.ply`
- Convert the `.ply` to a `.splat` using `GaussianSplats3D`'s tool
- Upload the `.splat` to the `GaussianSplats3D` Viewer

The screenshot included is what I have in `Super Splash`. I added a plane – which I believe is set at the center of the scene, right? So that should give me some idea, how far off my splat is. And the video here:
https://dl.rowild.at/GaussianSplats3D_upside_down.mp4
shows what happens in the `GaussianSplat3D` App.

I know that it is possible (and even necessary) to set the camera appropriately, when it comes to distance (the camera is way too close in `GaussianSplat3d`- but is it too close because of a different scaling or "simply" because of the position of the camera?). However, when it comes to the y-axis (I think that is the axis that goes from "center to bottom / top" in Three.js) is completely turned around. And here is, where I wonder, if there is a common "rule" as to what is up and what is down. I think I would understand when the splat would be turned around 90 degrees (that would be an indicator for the different definitions of the z-axis), but this does not seem to be the case here.

Is there anything I am misunderstanding here?

![Screenshot 2023-11-06 at 17 47 22](https://github.com/playcanvas/super-splat/assets/213803/3a8fe099-4e67-42ac-8211-6c25b6b61e7d)


If I may add my 2 cents of an idea of what I would love to see in Super Splash (I am a mediocre Blender user): it would be great to have an "automatic center" feature. Maybe is is possible to calculate the bounding box of a splat, so that a feature like blender's "Geometry to Center" could be applied – "Splat to Center" e.g..

And, after having centered my splat and further manipulated the rotation and fine-tuned the xyz-position of my splat, then I would like to be able to click a button that says "Apply changes". That would keep the current positioning of the splat, but re-caluculate all items (splats?) according to their new position in the `Super Splash` world. 

I think those 2 features would provide a solid starting point for exporting. I would know what I have to change in my destination software once I know how the splat is orientated in `Super Splat`.

But my thinking is very much dominated by Blender and by meshes and I am aware that I might be totally wrong with all that.

**Comment by slimbuck** [comment]
Hi @rowild,

I went ahead and added .splat export option to the editor. We don't support loading yet, we may still add this.

I also added really simple positioning control. You can toggle the scene center axis and set rotation/position now.

Please give that a go and let me know if it works ok for you.

Thanks!

**Comment by rowild** [comment]
Hi, @slimbuck!

This is pretty impressive: getting ones features ideas fulfilled! Like Christmas! Wow! Thank you very, very much!

I do like the axes and the option to move the model! I think it will also be a time saver to have a `.splat` exporter (haven't used it yet). What does the splat exporter do different from the other (aside from data formatting)? Does it handle the up/down axis differently? 

I do have some problems, though; I include a video that showcase my shortcomings:
https://dl.rowild.at/super-splash-3d-positioning-problem.mp4

It seems that when I try to rotate the model from left to right, the model actually rotates around the y-axes (the top-to-botton axis). But what I try to do is to turn the model i a manner, so that I can bring the backside to the front (in order to clean it up).
Also, when dragging the mouse up and down, the model flips, but only up to a certain point, where it then remains unresponsive. Is there a constraint anywhere in the "3D world" rotation that prohibits me from rotate an object any farther?

I hope I expressed myself clear enough – if not please let me know, I'll give it another try.

Thank you very much again! This is really, really great!

Also: Which origin is shown, when I click "Show origin"? The model's center? Or the world center? Because, again, when I rotate the object, the axes rotate along, and it does not seem that the axes's center is placed at the world's center – which is where? Is there a chance to 
- show the world center?
- move/re-place the model relative to the world center? (Because that is not what is happening now, right? Or am I mistaken?)

**Comment by slimbuck** [comment]
Hi @rowild,

The axis shows the world's origin and axes.

When you left click the viewport and drag, you are orbiting the camera around the object (not changing the orientation of the object itself). There is limit at the north and south poles, which is where the camera stops.

In order to change the rock's orientation, you must modify the "Rotation" entry on the left panel.

It looks like the rock probably needs a -90 degree rotation around the X-axis:
<img width="319" alt="Screenshot 2023-11-08 at 21 42 08" src="https://github.com/playcanvas/super-splat/assets/11276292/d8c65ce8-5ff4-4932-982b-cdf7ac473ccd">

Give that a try and also adjust the Y and Z rotation to see the object rotate around the green/blue axis.

Thanks!

**Reply by rowild** [reply]
It seems I did not react to your last entry from a long time ago! Sorry for that!
Your help was and is much appreciated!

**Comment by tksharpless** [comment]
Splatting tools tend to generate arbitrary coordinate systems, while many downstream applications need or would benefit from standardized coordinates,  As the premier splat editor, SuperSplat should provide better support for that.

It would be very helpful if you could add to the User Guide a page on coordinate systems, with explicit instructions how to rescale rotate and recenter the internal coordinates of a splat, and to verify that those changes reach exported files -- or else a statement of why this is not possible.  And it would be great if the UI could optionally display tripods showing the origin and orientation of each splat's internal coordinates.
